workflow/controller/operator.go

```diff
@@ -2126,6 +2121,48 @@ func (woc *wfOperationCtx) onNodeComplete(node *wfv1.NodeStatus) {
 	)
 }
 
+// recordNodePhaseChangeEvents creates WorkflowNode Kubernetes events for each node
+// that has changes logged during this execution of the operator loop.
+func (woc *wfOperationCtx) recordNodePhaseChangeEvents(old *wfv1.Nodes, new *wfv1.Nodes) {
+	if !woc.controller.Config.NodeEvents.IsEnabled() {
+		return
+	}
+	oldNodes := *old
+	newNodes := *new
+
+	// Check for newly added nodes; send an event for new nodes
+	for nodeName, newNode := range newNodes {
+		_, ok := oldNodes[nodeName]
+		if !ok {
+			if newNode.Phase == wfv1.NodeRunning {
+				woc.recordNodePhaseEvent(&newNode)
+			} else if newNode.Completed() {
+				ephemeralNode := newNode.DeepCopy()
```

Comment:
I make a copy because the next line modifies the phase. The phase I want to report is different from the node phase.

Reply:
Changed to mutate existing node instead of making a copy in https://github.com/argoproj/argo-workflows/pull/5531/commits/1818a707e17573cb519d1943334c0e9f050542f1